Enumerable-methods

bertil291utn opened the pull request:
## Enumerable methods

In this project, I have written most of the enumerable Ruby methods
- My each
- My each with index
- My select
- My all
- My any
- My none
- My count
- My map
- My inject

I  expect your feedback, thank you :computer: :rocket:

**UPDATE:**
Stickler is given an error about the rubocop linter. As it is in the requirement it is mandatory to disable this rubocop rule.

![image](https://user-images.githubusercontent.com/24902525/78963782-2157d280-7abe-11ea-991a-08546fc5988d.png)
![image](https://user-images.githubusercontent.com/24902525/78963842-61b75080-7abe-11ea-98b3-0d8ee02c477c.png)

But rubocop itself it says that is unnecessary disabling
`W: Lint/RedundantCopDisableDirective: Unnecessary disabling of Style/CaseEquality `

**Should I remove the rule disable in my file?**

bertil291utn committed:
My each method
- Add my each method clone to each ruby method

bertil291utn committed:
My each with index method
- Add my each  with index method clone to each with index
 ruby method

bertil291utn committed:
My select, all,any, none and count methods
- Clone these methods from ruby method

bertil291utn committed:
My map and my inject methods
- Add two methods clone of Ruby methods

bertil291utn committed:
Testing my_inject method
- Creating a method called #multiply_els which multiplies all the elements
of the array together by using #my_inject

bertil291utn committed:
Update README file

bertil291utn committed:
Update README file
- Update file name in description getting start

whiz25 reviewed (changes requested):
## Nice Progress :+1: 
Good progress with project. Your code covers most edge cases but we still have to fix a few cases before we can merge.
Check my comments for suggestions on how to fix the failing edge cases.

### Martin Afani

whiz25 commented on enumerable_method_asignment.rb:
```diff
@@ -0,0 +1,94 @@
+
+# method start
+module Enumerable
+  def my_each
+    index = 0
+    while index < size
+      yield(self[index])
+      index += 1
+    end
+  end
```
- [x] should be identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] my_each returns an Enumerator if no block is given
```ruby
array = Array.new(100) { rand(0...9) } }
array.my_each) == Enumerator #1 true
```

whiz25 commented on enumerable_method_asignment.rb:
```diff
@@ -0,0 +1,94 @@
+    end
+  end
+
+  def my_each_with_index
+    index = 0
+    while index < size
+      elem = self[index]
+      yield(elem, index)
+      index += 1
+    end
+  end
```
- [x] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] calls the given block once for each element in self
```ruby 
array = Array.new(100){rand(0...9)}
array.my_each_with_index == Enumerator #true
```

whiz25 commented on enumerable_method_asignment.rb:
```diff
@@ -0,0 +1,94 @@
+    end
+  end
+
+  def my_select
+    var_seleccionado = []
+    my_each { |elem| var_seleccionado.push(elem) if yield(elem) }
+    var_seleccionado
+  end
```
- [x] should be identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns an enumerator if no block is given
```ruby
array = Array.new(100){rand(0...9)}
array.my_select == Enumerator #true
```

whiz25 commented on enumerable_method_asignment.rb:
```diff
@@ -0,0 +1,94 @@
+    var_seleccionado
+  end
+
+  def my_all?
+    if !block_given?
+      my_each { |elem| return false if elem.nil? || !elem }
+    else
+      my_each { |elem| return false unless yield(elem) }
+    end
+    true
+  end
```
- [x] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
array = Array.new(100){rand(0...9)}
array.my_all?(Integer) == array.all?(Integer)
```
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
words = %w[dog door rod blade]
words.my_all?(/d/) == words.all?(/d/)
```
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
array.my_all?(3) == array.all?(3) #true
```

whiz25 commented on enumerable_method_asignment.rb:
```diff
@@ -0,0 +1,94 @@
+    true
+  end
+
+  def my_any?
+    if !block_given?
+      my_each { |elem| return true unless elem.nil? || !elem }
+    else
+      my_each { |elem| return true if yield(elem) }
+    end
+    false
+  end
```
- [x] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x]  when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
words = %w[dog door rod blade]
words.my_any?(Integer) == words.any?(Integer) #true
```
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
words.my_any?(/z/) == words.any?(/z/) #true
```
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
words.my_any?('cat') == words.any?('cat') #true
```

whiz25 commented on enumerable_method_asignment.rb:
```diff
@@ -0,0 +1,94 @@
+    false
+  end
+
+  def my_none?
+    my_each { |elem| return false if yield(elem) }
+    true
+  end
```
- [x] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
array = Array.new(100) { rand(0...9) }
array.my_none?(String) == true #true
```
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
words = %w[dog door rod blade]
words.my_none?(5) == words.none?(5) #true
```
- [x] Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
 words.my_none?(/z/) == words.none?(/z/) #true
 words.my_none?(/d/) == words.none?(/d/) #true
```
- [x] when no block or argument is given
```ruby
true_array = [nil, false, true, []] 
false_array = [nil, false, nil, false]
false_array.my_none? == true #true
true_array.my_none? == false #false
```

whiz25 commented on enumerable_method_asignment.rb:
```diff
@@ -0,0 +1,94 @@
+    count
+  end
+
+  def my_map(procedure = nil)
+    returned_array = []
+    my_each do |elem|
+      returned_array.push((procedure ? procedure.call(elem) : yield(elem)))
+    end
+    returned_array
+  end
```
- [x] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns an Enumerator if no block is given
```ruby
array = Array.new(100){rand(0...9)}
array.my_map == Enumerator #true
```

whiz25 commented on enumerable_method_asignment.rb:
```diff
@@ -0,0 +1,94 @@
+    returned_array
+  end
+
+  def my_inject(inicio = nil)
+    if inicio.nil?
+      resultado = self[0]
+      index = 1
+    else
+      resultado = inicio
+      index = 0
+    end
+
+    self[index..-1].my_each { |elem| resultado = yield(resultado, elem) }
+    resultado
+  end
+end
```
- [x] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject(:+) == array.inject(:+) #true
```
- [x] combines all elements of enum by applying a binary operation, specified by a block
```ruby
operation = proc { |sum, n| sum + n } 
search = proc { |memo, word| memo.length > word.length ? memo : word } 
array = Array.new(100){rand(0...9)}
words = %w[dog door rod blade]

array.my_inject(&operation) == array.inject(&operation) #true
actual = range.my_inject(4) { |prod, n| prod * n }
expected = range.inject(4) { |prod, n| prod * n }
actual == expected #true
words.my_inject(&search) == words.inject(&search) #true
```

bertil291utn committed:
FIxing first code review
- Add return enumerator in my_each,my_each_with_index,my_select and
my_map methods
-  Add to check when enters a integer, module and a regexp in my_all?,
my_any? and my_none? methods
- Accept symbol and procedures as argument, also a range
as enter data my_inject method
-  Disabling rubocop cops

maelfosso reviewed (changes requested):
# Changes required :red_circle: 

Hello,
Thank you for the improvements you've done.
However, there is still one bug to fix
```ruby
2.6.5 :004 > range = Range.new(5, 50)
 => 5..50 
2.6.5 :005 > range.my_inject(2, :*)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        6: from /home/maelfosso/.rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        5: from /home/maelfosso/.rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        4: from (irb):5
        3: from enumerable_method_asignment.rb:115:in `my_inject'
        2: from enumerable_method_asignment.rb:12:in `my_each'
        1: from enumerable_method_asignment.rb:116:in `block in my_inject'
LocalJumpError (no block given (yield))
2.6.5 :006 > range.inject(2, :*)
 => 2534507766809448170301050680505397403698136797413376000000000000 
```

Kindly do that change and submit a new code review request.
Thanks

Mael FOSSO

bertil291utn committed:
Fix second code review
- My_inject method add to receive as args a integer then a symbol
example: my_inject(2,:+)

mosesogwo reviewed (approved):
**Hi @bertil291utn:clap:. You are fit to :rocket:**

You can now merge your PR.

**Cheers:stars:**

### Status: Approved :heavy_check_mark:
